### PR TITLE
Fixes rescue hooks, adds a unit test to help prevent it from breaking again

### LINKED ISSUE #23763: Chasm Rescue Fishing Broken
Reporting client version: 514.1589

## Round ID:
9433

## Testmerges:
- [Adds the ability to make calls from outside the server to modify player ranks live](https://github.com/Skyrat-SS13/Skyrat-tg/pull/23521)
- [[TESTMERGE ME BEFORE MERGING] Armadyne Antitrust Act 2563 :: Disintegrates armadyne as a concept and replaces basically everything related](https://github.com/Skyrat-SS13/Skyrat-tg/pull/23424)

## Reproduction:
Fishing chasms has not been working for a few weeks. The runetext comes up that "you have fished out John Doe", but they dont actually come up. Has happened to at least a half dozen people that I've seen, using the rescue hook

### PR DESCRIPTION
**Pull request by vinylspiders**
## About The Pull Request

Fixes https://github.com/Skyrat-SS13/Skyrat-tg/issues/23763

Somewhere along https://github.com/tgstation/tgstation/pull/77739 and the following fishing PR's this feature got overlooked and broken. 

The args for `dispense_reward()` and `find_chasm_contents()` needed to be updated as they were just wrong.

This sets them straight and adds an additional fishing unit test for the rescue hook to hopefully prevent this edge case from being overlooked.

## Why It's Good For The Game

Rescue hooks work again, hooray.

## Changelog

:cl:
fix: rescue hooks will once again drop the mob next to the fisherman instead of just displaying a balloon alert and doing nothing
/:cl:

### CODE REVIEW

File: code/modules/fishing/sources/_fish_source.dm [hunk 1]
```diff
@@ -132,7 +132,11 @@ GLOBAL_LIST_INIT(preset_fish_sources, init_subtypes_w_path_keys(/datum/fish_sour
 		return
 	if(isitem(reward)) //Try to put it in hand
 		INVOKE_ASYNC(fisherman, TYPE_PROC_REF(/mob, put_in_hands), reward)
-	fisherman.balloon_alert(fisherman, "caught [reward]!")
+		fisherman.balloon_alert(fisherman, "caught [reward]!")
+	else // for fishing things like corpses, move them to the turf of the fisherman
+		INVOKE_ASYNC(reward, TYPE_PROC_REF(/atom/movable, forceMove), get_turf(fisherman))
+		fisherman.balloon_alert(fisherman, "caught something!")
```
Comment (Ghommie) [inline code/modules/fishing/sources/_fish_source.dm:138]:
```suggestion
	else // for fishing things like corpses, move them to the turf of the fisherman
		INVOKE_ASYNC(reward, TYPE_PROC_REF(/atom/movable, forceMove), get_turf(fisherman))
	fisherman.balloon_alert(fisherman, "caught [reward]!")
```

File: code/modules/unit_tests/fish_unit_tests.dm [hunk 2]
```diff
@@ -123,5 +123,57 @@
...
+	the_hole.drop(no_brain)
+	the_hole.drop(empty)
+	the_hole.drop(dummy)
+
+	// pretend like this mob has a mind. they should be fished up first
+	get_in_the_hole.mind = TRUE
+
+	SEND_SIGNAL(the_hole, COMSIG_PRE_FISHING) // we need to do this for the fishing spot component to be attached
+	var/datum/component/fishing_spot/the_hole_fishing_spot = the_hole.GetComponent(/datum/component/fishing_spot)
+	var/datum/fish_source/fishing_source = the_hole_fishing_spot.fish_source
```
Comment (Ghommie) [inline code/modules/unit_tests/fish_unit_tests.dm:157]: To be frank, you could simply use any preset fishing spot like this:
```
	var/datum/fish_source/fishing_source = GLOB.preset_fish_sources[/datum/fish_source/chasm]
```
but I understand why you would go through the long way.

**Review by Ghommie — CHANGES_REQUESTED** [review]
I've no idea why the unit test is failing.

File: code/modules/unit_tests/fish_unit_tests.dm [hunk 3]
```diff
@@ -123,5 +123,64 @@
...
+// we want no default spawns in this unit test
+/datum/chasm_detritus/restricted/bodies/no_defaults
+	default_contents_chance = 0
+
+/// Checks that we are able to fish people out of chasms with priority and that they end up in the right location
+/datum/unit_test/fish_rescue_hook
+
+/datum/unit_test/fish_rescue_hook/Run()
+	var/mob/living/carbon/human/consistent/get_in_the_hole = allocate(/mob/living/carbon/human/consistent) // create a human dummy to drop in the chasm
+	var/mob/living/basic/mining/lobstrosity/consistent/you_too = allocate(/mob/living/basic/mining/lobstrosity) // create some mindless mobs to fill the contents a bit
```
Comment (Ghommie) [inline code/modules/unit_tests/fish_unit_tests.dm:135]:
```suggestion
	var/mob/living/basic/mining/lobstrosity/you_too = allocate(/mob/living/basic/mining/lobstrosity) // create some mindless mobs to fill the contents a bit
```

File: code/modules/fishing/fish/chasm_detritus.dm [hunk 4]
```diff
@@ -40,30 +40,28 @@ GLOBAL_LIST_INIT_TYPED(chasm_detritus_types, /datum/chasm_detritus, init_chasm_d
 		),
 	)
 
-/datum/chasm_detritus/proc/dispense_reward(turf/fishing_spot, turf/fisher_turf)
-	if (prob(default_contents_chance))
-		var/default_spawn = pick(default_contents[default_contents_key])
-		return new default_spawn(fisher_turf)
-	return find_chasm_contents(fishing_spot, fisher_turf)
+/datum/chasm_detritus/proc/dispense_reward(reward_path, mob/fisherman, turf/fishing_spot)
```
[the remark this answers is not in the document]
Comment (Ghommie) [reply code/modules/fishing/fish/chasm_detritus.dm:43]: This arg is inconsequential, unused and, because the reward path is the same as the type of this datum, redundant.

To be clear, that proc you're showing me in that snippet of code is actually
`/datum/fish_source/proc/dispense_reward()`, while this is `/datum/chasm_detritus/proc/dispense_reward()`. Same name, different procpath.

File: code/modules/fishing/sources/_fish_source.dm [hunk 5]
```diff
@@ -132,7 +132,11 @@ GLOBAL_LIST_INIT(preset_fish_sources, init_subtypes_w_path_keys(/datum/fish_sour
...
+
```
[the remark this answers is not in the document]
Comment (Ghommie) [reply code/modules/fishing/sources/_fish_source.dm:139]: I prefer it to be explicit, also because the reward is likely to appear partly hidden by the player mob at first, as most movables have a lower layer than mobs. But I won't press on.

File: code/modules/fishing/sources/_fish_source.dm [hunk 6]
```diff
@@ -141,7 +145,7 @@ GLOBAL_LIST_INIT(preset_fish_sources, init_subtypes_w_path_keys(/datum/fish_sour
 	if(reward_path == FISHING_DUD)
 		return
 	if(ispath(reward_path, /datum/chasm_detritus))
-		return GLOB.chasm_detritus_types[reward_path].dispense_reward(fishing_spot, get_turf(fisherman))
+		return GLOB.chasm_detritus_types[reward_path].dispense_reward(reward_path, fisherman, fishing_spot)
```
[the remark this answers is not in the document]
Comment (Ghommie) [reply code/modules/fishing/sources/_fish_source.dm:148]: Again, don't confuse this with  `/datum/fish_source/proc/spawn_reward(reward_path, mob/fisherman, turf/fishing_spot)`

**Review by Ghommie — CHANGES_REQUESTED** [review]

File: code/modules/fishing/fish/chasm_detritus.dm (hunk 4 above)
[the remark this answers is not in the document]
Comment (Ghommie) [reply code/modules/fishing/fish/chasm_detritus.dm]: sure, i'm in for a different name, to make it less puzzling.

**Review by Ghommie — APPROVED** [review]

**Comment by distributivgesetz** [comment]
and i thought i was going crazy when failing to save people as a paramed
